docs/site/Interceptors.md

```diff
@@ -0,0 +1,651 @@
+- Override the method invocation, for example, return from cache
+
+The following diagram illustrates how interceptors can be applied to the
+invocation of a method on the controller class.
+
+![Interceptors](imgs/interceptors.png)
+
+## Basic use
+
+### Interceptors on controllers
+
+Interceptors are supported for public controller methods (including both static
+and prototype) and handler functions for REST routes.
+
+Controller methods decorated with `@intercept` are invoked with applied
+interceptors for corresponding routes upon API requests.
+
+```ts
+import {intercept} from '@loopback/context';
+
+@intercept(log) // `log` is an interceptor function
+export class OrderController {
+  @intercept('caching-interceptor') // `caching-interceptor` is a binding key
+  async listOrders(userId: string) {
+    // ...
+  }
+}
+```
+
+**NOTE**: `log` and `'caching-interceptor'` are illustrated in
+[Example interceptors](#example-interceptors).
+
+It's also possible to configure global interceptors that are invoked before
+method level interceptors. For example, the following code registers a global
+`caching-interceptor` for all methods.
+
+```ts
+app
+  .bind('caching-interceptor')
+  .toProvider(CachingInterceptorProvider)
+  .apply(asInterceptor);
+```
+
+Global interceptors are also executed for route handler functions without a
+controller class. See an example in
+[Route Handler](Routes.md#using-partial-openapi-spec-fragments).
+
+### Create a proxy to apply interceptors
+
+A
+[proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
+can be created from the target class or object to apply interceptors. This is
+useful for the case that a controller declares dependencies of repositories or
+services and would like to allow repository or service methods to be
+intercepted.
+
+```ts
+import {createProxyWithInterceptors} from '@loopback/context';
+
+const proxy = createProxyWithInterceptors(controllerInstance, ctx);
+const msg = await proxy.greet('John');
+```
+
+There is also an `asProxyWithInterceptors` option for binding resolution or
+dependency injection to return a proxy for the class to apply interceptors when
+methods are invoked.
+
+```ts
+class DummyController {
+  constructor(
+    @inject('my-controller', {asProxyWithInterceptors: true})
+    public readonly myController: MyController,
+  ) {}
+}
+ctx.bind('my-controller').toClass(MyController);
+ctx.bind('dummy-controller').toClass(DummyController);
+const dummyController = await ctx.get<DummyController>('dummy-controller');
+const msg = await dummyController.myController.greet('John');
+);
+```
+
+Or:
+
+```ts
+const proxy = await ctx.get<MyController>('my-controller', {
+  asProxyWithInterceptors: true,
+});
+const msg = await proxy.greet('John');
+```
+
+Please note synchronous methods (which don't return `Promise`) are converted to
+be asynchronous in the proxy so that interceptors can be applied. For example,
```

Comment:
> Please note synchronous methods (which don't return `Promise`) are converted to
> be asynchronous in the proxy so that interceptors can be applied.

I think this is not entirely correct. IIUC, synchronous methods may be converted to asynchronous, this happen when there is at least one async interceptor. Because it's not possible to determine whether the proxied method will end up being sync or async, sync methods are converted into methods returning `ValueOrPromise`.

Could you please update the documentation to make the behavior more clear to users?